### PR TITLE
Fix #4042: Implement success criteria metrics for lesson checkpointing

### LINKED ISSUE #4042: Implement success criteria metrics for lesson checkpointing
**Is your feature request related to a problem? Please describe.**
Lesson checkpointing was implemented as a GSoC project last summer, but we didn't have time to implement the success criteria metrics expected for the feature [here](https://docs.google.com/document/d/1d8yjwz76mngtsPRxC7fubgLKg8mfA7kG1sWRWdbiaVw/edit#bookmark=id.2zyjd5vygmcv).

**Describe the solution you'd like**
The metrics should be implemented using the existing analytics framework. This feature should be documented at least in this issue with a high-level overview with the changes that need to be done in the code before actually starting implementation. A design document is not needed.

**Describe alternatives you've considered**
N/A

**Additional context**
None

### PR DESCRIPTION
**Pull request by theMr17**
## Explanation

Fixes #4042

This PR implements the success criteria metrics required for lesson checkpointing as described [here](https://docs.google.com/document/d/1d8yjwz76mngtsPRxC7fubgLKg8mfA7kG1sWRWdbiaVw/edit#bookmark=id.2zyjd5vygmcv). A high level overview of the implementation is documented on the issue thread [here](https://github.com/oppia/oppia-android/issues/4042#issuecomment-1929799546).

#### Changes to FakeAnalyticsEventLogger
- `fun getOldestEvents(count: Int): List<EventLog>` is necessary to retrieve a list of a pre-defined count of oldest events from all the logged events.
- `fun getLoggedEvent(predicate: (EventLog) -> Boolean): EventLog?` serves to acquire a reference to a logged event when the context of the event is known, but the event index is uncertain or potentially surrounded by other events. Thus, extracting the event via index might be difficult.

## Essential Checklist

- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## For UI-specific PRs only

If your PR includes UI-related changes, then:
- Add screenshots for portrait/landscape for both a tablet & phone of the before & after UI changes
- For the screenshots above, include both English and pseudo-localized (RTL) screenshots (see [RTL guide](https://github.com/oppia/oppia-android/wiki/RTL-Guidelines))
- Add a video showing the full UX flow with a screen reader enabled (see [accessibility guide](https://github.com/oppia/oppia-android/wiki/Accessibility-A11y-Guide))
- For PRs introducing new UI elements or color changes, both light and dark mode screenshots must be included
- Add a screenshot demonstrating that you ran affected Espresso tests locally & that they're passing

### CODE REVIEW

**Comment by theMr17** [comment]
@adhiamboperes PTAL. This PR is ready for a review.

File: app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivityPresenter.kt [hunk 1]
```diff
@@ -330,6 +332,7 @@ class ExplorationActivityPresenter @Inject constructor(
       showUnsavedExplorationDialogFragment()
       return
     }
+    learnerAnalyticsLogger.explorationAnalyticsLogger.value?.logLessonSavedAdvertently()
```
Comment (adhiamboperes) [inline app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivityPresenter.kt]: I believe this should be inside  `showDialogFragmentBasedOnCurrentCheckpointState()`, because it is what handles the back button behaviour when checkpointing is enabled.

The current placement of this call will cause the log to be created every time, even when checkpointing is disabled, as it is outside the if block.

File: app/src/sharedTest/java/org/oppia/android/app/player/exploration/ExplorationActivityTest.kt [hunk 2]
```diff
@@ -1932,6 +1937,129 @@ class ExplorationActivityTest {
...
+    testCoroutineDispatchers.runCurrent()
+
+    val lessonSavedAdvertentlyEventCount = fakeAnalyticsEventLogger.countEvents {
+      it.context.activityContextCase == LESSON_SAVED_ADVERTENTLY_CONTEXT
+    }
+    assertThat(lessonSavedAdvertentlyEventCount).isEqualTo(0)
+  }
+
+  @Test
+  fun testExpActivity_replayExploration_pressToolbarBackIcon_doesNotLogLessonSavedAdvertentlyEvent() { // ktlint-disable max-line-length
```
Comment (adhiamboperes) [inline app/src/sharedTest/java/org/oppia/android/app/player/exploration/ExplorationActivityTest.kt]: You can shorten the test name to fit.
```suggestion
 fun testExpActivity_replayExp_pressToolbarBackIcon_doesNotLogLessonSavedAdvertentlyEvent() {
```

File: app/src/sharedTest/java/org/oppia/android/app/player/exploration/ExplorationActivityTest.kt [hunk 3]
```diff
@@ -1932,6 +1937,129 @@ class ExplorationActivityTest {
...
+    // Click on 'X' icon on toolbar.
+    onView(withContentDescription(R.string.nav_app_bar_navigate_up_description)).perform(click())
+    testCoroutineDispatchers.runCurrent()
+
+    val lessonSavedAdvertentlyEventCount = fakeAnalyticsEventLogger.countEvents {
+      it.context.activityContextCase == LESSON_SAVED_ADVERTENTLY_CONTEXT
+    }
+    assertThat(lessonSavedAdvertentlyEventCount).isEqualTo(0)
+  }
+
```
Comment (adhiamboperes) [inline app/src/sharedTest/java/org/oppia/android/app/player/exploration/ExplorationActivityTest.kt:2062]: Could you please add tests to verify that the event is not logged when checkpointing is disabled?

In the test class, that should be created by setting  `shouldSavePartialProgress = false` when creating the activity intent.

File: domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressController.kt [hunk 4]
```diff
@@ -443,6 +443,8 @@ class ExplorationProgressController @Inject constructor(
                 ControllerState(
                   ExplorationProgress(),
                   message.isRestart,
+                  isResume = !message.isRestart &&
+                    message.explorationCheckpoint != ExplorationCheckpoint.getDefaultInstance(),
```
Comment (adhiamboperes) [inline domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressController.kt]: This is kind of difficult to understand, especially for someone that may not be very farmiliar with the codebase.

Adding a comment may be helpful.

File: testing/src/main/java/org/oppia/android/testing/FakeAnalyticsEventLogger.kt
Comment (adhiamboperes) [inline testing/src/main/java/org/oppia/android/testing/FakeAnalyticsEventLogger.kt:1]: Could you please explain in your PR why these new methods were needed?

File: utility/src/main/java/org/oppia/android/util/logging/EventBundleCreator.kt [hunk 5]
```diff
@@ -88,27 +114,6 @@ import org.oppia.android.util.platformparameter.PlatformParameterValue
 import java.util.concurrent.atomic.AtomicInteger
 import javax.inject.Inject
 import javax.inject.Singleton
-import org.oppia.android.app.model.EventLog.AbandonSurveyContext as AbandonSurveyEventContext
```
Comment (adhiamboperes) [inline utility/src/main/java/org/oppia/android/util/logging/EventBundleCreator.kt]: Please look into why these importas changed. I expected to only see additions of the new contexts.

File: domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt [hunk 6]
```diff
@@ -1967,8 +1971,8 @@ class ExplorationProgressControllerTest {
     waitForGetCurrentStateSuccessfulLoad()
 
     val exploration = loadExploration(TEST_EXPLORATION_ID_2)
-    val eventLog = fakeAnalyticsEventLogger.getMostRecentEvent()
-    assertThat(fakeAnalyticsEventLogger.getEventListCount()).isEqualTo(1)
+    val eventLog = fakeAnalyticsEventLogger.getOldestEvent()
```
Comment (adhiamboperes) [inline domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt:1974]: Why is this change needed? As I understand, no changes were made to  logging when an exploration is started, only when it is resumed.

File: domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt [hunk 7]
```diff
@@ -1986,8 +1990,8 @@ class ExplorationProgressControllerTest {
     waitForGetCurrentStateSuccessfulLoad()
 
     // Resuming shouldn't log a 'start card' event.
-    val eventLog = fakeAnalyticsEventLogger.getMostRecentEvent()
-    assertThat(fakeAnalyticsEventLogger.getEventListCount()).isEqualTo(1)
+    val eventLog = fakeAnalyticsEventLogger.getOldestEvent()
+    assertThat(fakeAnalyticsEventLogger.getEventListCount()).isEqualTo(4)
```
Comment (adhiamboperes) [inline domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt:1994]: Why are an additional 3 events being logged on resume?

File: domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt [hunk 8]
```diff
@@ -2003,8 +2007,8 @@ class ExplorationProgressControllerTest {
     restartExploration(TEST_TOPIC_ID_0, TEST_STORY_ID_0, TEST_EXPLORATION_ID_2)
     waitForGetCurrentStateSuccessfulLoad()
 
-    val (eventLog1, eventLog2) = fakeAnalyticsEventLogger.getMostRecentEvents(count = 2)
-    assertThat(fakeAnalyticsEventLogger.getEventListCount()).isEqualTo(2)
```
Comment (adhiamboperes) [inline domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt:2007]: No changes were made to start over, so why has the test changed?

File: domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt [hunk 9]
```diff
@@ -2092,7 +2096,7 @@ class ExplorationProgressControllerTest {
     val hasEngagementEvent = fakeAnalyticsEventLogger.hasEventLogged {
       it.context.activityContextCase == REACH_INVESTED_ENGAGEMENT
     }
-    val eventLog = fakeAnalyticsEventLogger.getMostRecentEvent()
+    val eventLog = fakeAnalyticsEventLogger.get2ndMostRecentEvent()
```
Comment (adhiamboperes) [inline domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt]: When a new exploration is played, which event is being recorded first to make this the second event?

File: domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt [hunk 10]
```diff
@@ -2309,17 +2420,20 @@ class ExplorationProgressControllerTest {
...
       hasSkillIdThat().isEqualTo("test_skill_id_0")
     }
+    assertThat(eventLogList[1]).hasProgressSavingSuccessContextThat {
+      containsTestExp2Details()
+    }
   }
 
   @Test
-  fun testHint_offered_logsHintOfferedEvent() {
+  fun testHint_offered_logsHintOfferedEvent_logsProgressSavingSuccessEvent() {
```
Comment (adhiamboperes) [inline domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt:2439]: For this and the hint/solution tests below, is checkpoint state changed? I do  not understand why the progress save success event is being logged otherwise.

**Review by adhiamboperes — CHANGES_REQUESTED** [review]
Thanks @theMr17!

I have left some questions inline.

Apologies for the delay in review!

File: app/src/sharedTest/java/org/oppia/android/app/player/exploration/ExplorationActivityTest.kt (hunk 2 above)
Comment (theMr17) [reply app/src/sharedTest/java/org/oppia/android/app/player/exploration/ExplorationActivityTest.kt]: Done.

File: utility/src/main/java/org/oppia/android/util/logging/EventBundleCreator.kt (hunk 5 above)
Comment (theMr17) [reply utility/src/main/java/org/oppia/android/util/logging/EventBundleCreator.kt]: Removed the unnecessary changes. However, this was done in an attempt to fix the lint failure in that file, as the imports were not in lexicographical order.

File: app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivityPresenter.kt (hunk 1 above)
Comment (theMr17) [reply app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivityPresenter.kt]: > The current placement of this call will cause the log to be created every time, even when checkpointing is disabled, as it is outside the if block.

The current placement of this call will not log the event when checkpointing is disabled because the if block just above it checks whether checkpointing is enabled or not. Incase checpointing is disabled the return statement in the if block will be executed and following lines including the log statement will not be executed.

But, you are right, this should be inside `showDialogFragmentBasedOnCurrentCheckpointState()` inside the `when` block to ensure that the event is logged only when lesson checkpoint is saved successfully.

File: app/src/sharedTest/java/org/oppia/android/app/player/exploration/ExplorationActivityTest.kt (hunk 3 above)
Comment (theMr17) [reply app/src/sharedTest/java/org/oppia/android/app/player/exploration/ExplorationActivityTest.kt:2062]: I have already added tests to verify that the event is not logged when checkpointing is disabled in the following tests.

1. [testExpActivity_replayExp_pressToolbarBackIcon_doesNotLogLessonSavedAdvertentlyEvent](https://github.com/oppia/oppia-android/blob/7e834553aaec240bae868350f401c9e5e34440c1/app/src/sharedTest/java/org/oppia/android/app/player/exploration/ExplorationActivityTest.kt#L2033)
2. [testExpActivity_replayExploration_pressBack_doesNotLogLessonSavedAdvertentlyEvent](https://github.com/oppia/oppia-android/blob/7e834553aaec240bae868350f401c9e5e34440c1/app/src/sharedTest/java/org/oppia/android/app/player/exploration/ExplorationActivityTest.kt#L2004C7-L2004C88)

File: domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressController.kt (hunk 4 above)
Comment (theMr17) [reply domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressController.kt]: I have added a comment. Also, I have removed `!message.isRestart` as it is not required here.

File: domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt (hunk 6 above)
Comment (theMr17) [reply domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt:1974]: When any exploration has finished loading, a checkpoint is saved from [here](https://github.com/oppia/oppia-android/blob/9e8553cdcebb5f40b170d33b2e02bbe02081530b/domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressController.kt#L894). So, this change was needed even when an exploration is started.

File: domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt (hunk 8 above)
Comment (theMr17) [reply domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt:2007]: This is similar to the previous [comment](https://github.com/oppia/oppia-android/pull/5336#discussion_r1553713388).

File: domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt (hunk 9 above)
Comment (theMr17) [reply domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt]: I would like to clarify that the "most recent event" here refers to the last logged event. So, the "2nd most recent event" means the second-to-last event. By that analogy, the last event logged here is the lesson progress saving success event. The event is logged due to the same reason as mentioned [here](https://github.com/oppia/oppia-android/pull/5336#discussion_r1556011900).

File: domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt (hunk 7 above)
Comment (theMr17) [reply domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt:1994]: The additional 3 events logged are progress saving success events. The event is logged 3 times because the checkpoint is saved 3 times in `beginExplorationImpl` at line 570 [here](https://github.com/oppia/oppia-android/blob/9e8553cdcebb5f40b170d33b2e02bbe02081530b/domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressController.kt#L565-L572).

File: domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt (hunk 10 above)
Comment (theMr17) [reply domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt:2439]: Whenever there is a change in hint state, for example a hint is unlocked or accessed, a corresponding checkpoint is also saved. So, a progress saving success event is logged.

File: testing/src/main/java/org/oppia/android/testing/FakeAnalyticsEventLogger.kt
Comment (theMr17) [reply testing/src/main/java/org/oppia/android/testing/FakeAnalyticsEventLogger.kt:1]: Done.

**Comment by theMr17** [comment]
@adhiamboperes, thanks for the review. I have addressed your comments, PTAL.

File: domain/src/main/java/org/oppia/android/domain/oppialogger/analytics/LearnerAnalyticsLogger.kt [hunk 11]
```diff
@@ -245,6 +245,31 @@ class LearnerAnalyticsLogger @Inject constructor(
...
+    /** Logs that correct answer was submitted in a resumed lesson. */
+    fun logResumeLessonSubmitCorrectAnswer() {
+      getExpectedStateLogger()?.logResumeLessonSubmitCorrectAnswer()
+    }
+
+    /** Logs that incorrect answer was submitted in a resumed lesson. */
+    fun logResumeLessonSubmitIncorrectAnswer() {
+      getExpectedStateLogger()?.logResumeLessonSubmitIncorrectAnswer()
+    }
+
```
Comment (adhiamboperes) [inline domain/src/main/java/org/oppia/android/domain/oppialogger/analytics/LearnerAnalyticsLogger.kt:272]: Tests for these are needed in the `LearnerAnalyticsLoggerTest`

File: domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressController.kt [hunk 12]
```diff
@@ -443,6 +443,18 @@ class ExplorationProgressController @Inject constructor(
...
+                   * The [message.explorationCheckpoint] is [ExplorationCheckpoint.getDefaultInstance()]
+                   * in the following 3 cases.
+                   *  - New exploration is started.
+                   *  - Saved Exploration is restarted.
+                   *  - Completed exploration is replayed.
+                   *
+                   * The [message.explorationCheckpoint] will contain the exploration checkpoint
+                   * only when a saved exploration is resumed.
+                   */
+                  isResume = message.explorationCheckpoint
```
Comment (adhiamboperes) [inline domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressController.kt]: This should just be a // comment like the one above it and not the multiline KDoC format. It is okay to be on multiple lines.

File: domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt (hunk 9 above)
Comment (adhiamboperes) [reply domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt]: Instead of creating a new `fun get2ndMostRecentEvent()` in FakeAnalyticsEventLogger, we can use the existing function, `  fun getMostRecentEvents(count: Int): List<EventLog>`. This will retur a list of the most recent `n` events. In this case, pass in 2, and get the latest 2 events.

` val eventLog = fakeAnalyticsEventLogger.getMostRecentEvents(2).first()
 `
 is semantically equal to 
 
` fun get2ndMostRecentEvent(): EventLog = getMostRecentEvents(count = 2)[0]`

And avoids the need to create a  `fun get3rdMostRecentEvent()` etc in future.

File: testing/src/main/java/org/oppia/android/testing/FakeAnalyticsEventLogger.kt [hunk 13]
```diff
@@ -30,6 +36,9 @@ class FakeAnalyticsEventLogger @Inject constructor() : AnalyticsEventLogger {
   /** Returns whether a certain event has been logged or not, based on the provided [predicate]. */
   fun hasEventLogged(predicate: (EventLog) -> Boolean): Boolean = eventList.find(predicate) != null
 
+  /** Returns a certain event if it has been logged or null, based on the provided [predicate]. */
+  fun getLoggedEvent(predicate: (EventLog) -> Boolean): EventLog? = eventList.find(predicate)
```
Comment (adhiamboperes) [inline testing/src/main/java/org/oppia/android/testing/FakeAnalyticsEventLogger.kt:37]: This is introducing a new pattern, but we already have a pattern for testing the event context via the EventLogSubject pattern. Please see my comment on the tests for how to address this.

File: domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt [hunk 14]
```diff
@@ -2372,16 +2502,29 @@ class ExplorationProgressControllerTest {
...
     submitMultipleChoiceAnswer(choiceIndex = 0)
 
-    val eventLog = fakeAnalyticsEventLogger.getMostRecentEvent()
-    assertThat(eventLog).hasHintUnlockedContextThat {
+    val hintOfferedEvent = fakeAnalyticsEventLogger.getLoggedEvent {
+      it.context.activityContextCase == HINT_UNLOCKED_CONTEXT
+    }.also {
+      assert(it != null)
+    }
+    assertThat(hintOfferedEvent!!).hasHintUnlockedContextThat {
```
Comment (adhiamboperes) [inline domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt:2513]: I think you should use the same approach here as in the 
`testHint_lastHintWithNoSol_offeredThenViewed_logsViewHintEvt_logsProgressSavingSuccessEvt()` below to retrieve the list of the last 2 events, of which the `hintOfferedEvent` will be the [0], and the `progressSavingSuccessEvent` will be the [1].

**Review by adhiamboperes — CHANGES_REQUESTED** [review]
Thanks @theMr17!

I have left follow up comments.

File: domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressController.kt (hunk 12 above)
Comment (theMr17) [reply domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressController.kt]: Done.

File: domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt (hunk 9 above)
Comment (theMr17) [reply domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt]: Removed the extra function and updated per suggestion.

File: testing/src/main/java/org/oppia/android/testing/FakeAnalyticsEventLogger.kt (hunk 13 above)
Comment (theMr17) [reply testing/src/main/java/org/oppia/android/testing/FakeAnalyticsEventLogger.kt:37]: You are right, this function is not required. I have removed it and followed the existing pattern.

File: domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt (hunk 14 above)
Comment (theMr17) [reply domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt:2513]: Done.

File: domain/src/main/java/org/oppia/android/domain/oppialogger/analytics/LearnerAnalyticsLogger.kt (hunk 11 above)
Comment (theMr17) [reply domain/src/main/java/org/oppia/android/domain/oppialogger/analytics/LearnerAnalyticsLogger.kt:272]: Added tests for these changes.

File: domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt (hunk 14 above)
Comment (theMr17) [reply domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt:2513]: The test is failing on robolectric after making the changes.

**Comment by theMr17** [comment]
@adhiamboperes PTAL.

File: domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt (hunk 14 above)
Comment (adhiamboperes) [reply domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt:2513]: Here is the stacktrace from CI:

```kotlin
org.oppia.android.domain.exploration.ExplorationProgressControllerTest > testHint_offered_logsHintOfferedEvent_logsProgressSavingSuccessEvent FAILED
    expected: HINT_UNLOCKED_CONTEXT
    but was : PROGRESS_SAVING_SUCCESS_CONTEXT       
```

From the test name, we want to assert that:
a) hint offerred event is logged
b) progress saving success event is logged

However, the test setup isn't done in such a way to allow this:

Can you try the following:

   ```
 val eventLogList = fakeAnalyticsEventLogger.getMostRecentEvents(2)

    val hintOfferedEvent = eventLogList[0])
    // use eventlogsubject to assert event details
   
 val progressSavingSuccessEvent = eventLogList[1])
 // use eventlogsubject to assert event details
```

**Review by adhiamboperes — CHANGES_REQUESTED** [review]
Thanks @theMr17!

I have left a comment on the failing eventlog test, please let me know if I need to clarify it further.

File: domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt (hunk 14 above)
Comment (theMr17) [reply domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt:2513]: > Can you try the following:
> 
> ```
> val eventLogList = fakeAnalyticsEventLogger.getMostRecentEvents(2)
> 
>  val hintOfferedEvent = eventLogList[0])
>  // use eventlogsubject to assert event details
> 
> val progressSavingSuccessEvent = eventLogList[1])
> // use eventlogsubject to assert event details
> ```

I have already made those changes. PTAL at the code [here](https://github.com/oppia/oppia-android/blob/bdf303493987e84eef49c3e60b3e43dfa67bb58d/domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt#L2447-L2455). The test is failing on Robolectric after making the suggested changes. Although, the test is passing locally.

I think the test is failing maybe because the order of the events in which they are logged is changing. This was the reason why I had introduced `fun getLoggedEvent(predicate: (EventLog) -> Boolean): EventLog?` function.

Comment (adhiamboperes) [reply domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt:2513]: In this case I think it may be okay to add the getLoggedEvent functio.

Comment (theMr17) [reply domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt:2513]: Done.

**Comment by theMr17** [comment]
@adhiamboperes PTAL. I have reintroduced the `getLoggedEvent(predicate: (EventLog) -> Boolean): EventLog?` function. I have also included tests for the new functions added in `FakeAnalyticsEventLogger`.

**Review by adhiamboperes — APPROVED** [review]
Thanks @theMr17! This LGTM!
